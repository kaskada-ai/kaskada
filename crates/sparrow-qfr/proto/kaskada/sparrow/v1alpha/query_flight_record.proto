syntax = "proto3";
package kaskada.sparrow.v1alpha;

import "google/protobuf/wrappers.proto";

option go_package = "gitlab.com/kaskada/kaskada/wren/gen/go/kaskada/sparrow/v1alpha";
option java_multiple_files = true;
option java_package = "com.kaskada.sparrow.v1alpha1";

// Metadata added at the beginning of the flight record file.
//
// Contains metadata on the query, plan, etc. useful for interepreting
// the flight record.
message FlightRecordHeader {
  // The version of the flight record format contained here.
  uint32 version = 1;

  // The request ID this flight record is for.
  string request_id = 2;

  // Build information about the Sparrow Binary.
  BuildInfo sparrow_build_info = 3;

  // Register activities that may be reported during the query.
  //
  // This includes the set of all activities which may be produced by any query.
  // Not all activities will be reported by any specific query.
  repeated RegisterActivity activities = 4;

  // Register metrics that may be reported during the query.
  //
  // Metrics may be reported as part of an activity (see ReportActivity) or
  // at the top level during execution (see ReportMetrics).
  //
  // The set of metrics are static within the query engine, and correspond
  // to any value that may be reported during the execution of a query. Not
  // all metrics need to be reported for any given query.
  repeated RegisterMetric metrics = 5;

  message BuildInfo {
    string sparrow_version = 1;
    string github_ref = 2;
    string github_sha = 3;
    string github_workflow = 4;
  }

  message RegisterActivity {
    uint32 activity_id = 1;

    string label = 2;

    // The ID of the parent activity.
    //
    // May be `None` / `null` to indicate a root activity.
    google.protobuf.UInt32Value parent_activity_id = 3;
  }

  message RegisterMetric {
    uint32 metric_id = 1;
    string label = 2;
    MetricKind kind = 3;

    // string description = 4;
    // string units = 5;

    enum MetricKind {
      METRIC_KIND_UNSPECIFIED = 0;

      METRIC_KIND_U64_GAUGE = 10;
      METRIC_KIND_I64_GAUGE = 11;
      METRIC_KIND_F64_GAUGE = 12;

      METRIC_KIND_U64_COUNTER = 20;
      METRIC_KIND_I64_COUNTER = 21;
      METRIC_KIND_F64_COUNTER = 22;
    }
  }
}

message FlightRecord {
  oneof record {
    // Register an execution thread.
    // This is sent for each computation thread spawned for a specific query.
    RegisterThread register_thread = 1;

    /// Report an activity execution during the query.
    ReportActivity report_activity = 2;

    /// Report a top-level metric.
    ReportMetrics report_metrics = 3;
  }

  message RegisterThread {
    uint32 thread_id = 1;
    string label = 2;
  }

  message ReportActivity {
    uint32 activity_id = 1;
    uint32 thread_id = 2;

    // The time within the trace at which this event was recorded.
    uint64 wall_timestamp_us = 3;

    // Wall clock time spent executing this event.
    //
    // This should be `0` for point-wise events.
    uint64 wall_duration_us = 4;

    // CPU time spent executing this event.
    uint64 cpu_duration_us = 5;

    // Metrics to report associated with the given activity.
    repeated MetricValue metrics = 6;
  }

  message ReportMetrics {
    uint32 thread_id = 1;

    // The time within the trace at which this event was recorded.
    uint64 wall_timestamp_us = 2;

    // Metrics to report against the given thread.
    repeated MetricValue metrics = 3;
  }
}

message MetricValue {
  uint32 metric_id = 1;

  oneof value {
    uint64 u64_value = 2;
    int64 i64_value = 3;
    double f64_value = 4;
  }
}