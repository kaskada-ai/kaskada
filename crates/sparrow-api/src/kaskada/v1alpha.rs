#![allow(clippy::large_enum_variant)] // Could box the pulsar subscription, but for now just allowing this.
#![allow(clippy::derive_partial_eq_without_eq)]
use std::str::FromStr; // stop clippy erroring on generated code from tonic (proto)

use itertools::Itertools;
use url::Url;

use crate::kaskada::v1alpha::object_store_destination::ResultPaths;
use crate::kaskada::v1alpha::operation_plan::tick_operation::TickBehavior;

tonic::include_proto!("kaskada.kaskada.v1alpha");

/// Traits for the `Slice`
mod slice_traits;

/// Implementations and traits for parts of the plan
mod plan_impl;

/// Traits for [PreparedFile]
mod prepared_file_impl;

/// Traits for [Schema]
mod schema_traits;

impl ComputeTable {
    pub fn name(&self) -> &str {
        &self.config().name
    }

    pub fn config(&self) -> &TableConfig {
        self.config.as_ref().expect("tables should have a config")
    }
}

impl ExecuteResponse {
    /// Returns an owned vec of the output paths for this execution.
    //
    /// Returns `None` if not an `ObjectStoreOutput`.
    pub fn output_paths(&self) -> Option<Vec<String>> {
        match &self.destination {
            Some(output) => match &output.destination {
                Some(destination::Destination::ObjectStore(store)) => store
                    .output_paths
                    .as_ref()
                    .map(|output_paths| output_paths.paths.clone()),
                Some(_) | None => None,
            },
            None => None,
        }
    }

    /// Returns a mutable reference to the output paths.
    ///
    /// Returns `None` if not an `ObjectStoreOutput`.
    pub fn output_paths_mut(&mut self) -> Option<&mut ResultPaths> {
        match &mut self.destination {
            Some(output) => match &mut output.destination {
                Some(destination::Destination::ObjectStore(store)) => match &mut store.output_paths
                {
                    Some(output_path) => Some(output_path),
                    None => None,
                },
                Some(_) | None => None,
            },
            None => None,
        }
    }
}

impl FeatureSet {
    pub fn new(query: &str, formulas: Vec<(&str, &str)>) -> Self {
        Self {
            formulas: formulas
                .into_iter()
                .enumerate()
                .map(|(index, (name, formula))| Formula {
                    name: name.to_owned(),
                    formula: formula.to_owned(),
                    source_location: format!("FeatureSet::new formula {index}"),
                })
                .collect(),
            query: query.to_owned(),
        }
    }
}

impl TableConfig {
    /// Creates a table config with the given parameters and a `KaskadaSource`,
    /// which represents a file-backed table.
    pub fn new_with_table_source(
        name: &str,
        uuid: &uuid::Uuid,
        time_column_name: &str,
        subsort_column_name: Option<&str>,
        group_column_name: &str,
        grouping: &str,
    ) -> Self {
        Self {
            name: name.to_owned(),
            uuid: uuid.to_string(),
            time_column_name: time_column_name.to_owned(),
            subsort_column_name: subsort_column_name.map(|s| s.to_owned()),
            group_column_name: group_column_name.to_owned(),
            grouping: grouping.to_owned(),
            source: Some(Source {
                source: Some(source::Source::Kaskada(KaskadaSource {})),
            }),
        }
    }

    pub fn new2(
        name: &str,
        time_column_name: &str,
        subsort_column_name: Option<&str>,
        group_column_name: &str,
        grouping: &str,
    ) -> Self {
        Self::new_with_table_source(
            name,
            &uuid::Uuid::new_v4(),
            time_column_name,
            subsort_column_name,
            group_column_name,
            grouping,
        )
    }

    pub fn with_table_source(self) -> Self {
        Self::new_with_table_source(
            &self.name,
            &uuid::Uuid::from_str(&self.uuid).unwrap(),
            &self.time_column_name,
            self.subsort_column_name.as_deref(),
            &self.group_column_name,
            &self.grouping,
        )
    }
}

impl Formula {
    pub fn new(name: &str, formula: &str) -> Self {
        Self {
            name: name.to_owned(),
            formula: formula.to_owned(),
            source_location: name.to_owned(),
        }
    }
}

// Display PlanHash as the upper hex encoded vector.
impl std::fmt::Display for PlanHash {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            self.hash
                .iter()
                .format_with("", |elt, f| f(&format_args!("{elt:X}")))
        )
    }
}

impl SourceData {
    pub fn try_from_local(path: &std::path::Path) -> anyhow::Result<source_data::Source> {
        let path = match path.extension().and_then(|ext| ext.to_str()) {
            Some("parquet") => {
                source_data::Source::ParquetPath(format!("file://{}", path.display()))
            }
            Some("csv") => source_data::Source::CsvPath(format!("file://{}", path.display())),
            unsupported => anyhow::bail!("Unsupported extension {:?}", unsupported),
        };
        Ok(path)
    }

    pub fn try_from_url(url: &Url, extension: Option<&str>) -> anyhow::Result<source_data::Source> {
        let path = match extension {
            Some("parquet") => source_data::Source::ParquetPath(url.to_string()),
            Some("csv") => source_data::Source::CsvPath(url.to_string()),
            unsupported => anyhow::bail!("Unsupported extension {:?}", unsupported),
        };
        Ok(path)
    }
}

impl std::fmt::Display for FenlDiagnostics {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            Itertools::intersperse(
                self.fenl_diagnostics
                    .iter()
                    .map(|d| -> &str { &d.formatted }),
                ""
            )
            .flat_map(|f| f.lines())
            .map(|s| s.trim_end())
            .format("\n")
        )
    }
}

impl std::fmt::Display for TickBehavior {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TickBehavior::Minutely => write!(f, "minutely"),
            TickBehavior::Hourly => write!(f, "hourly"),
            TickBehavior::Daily => write!(f, "daily"),
            TickBehavior::Monthly => write!(f, "monthly"),
            TickBehavior::Yearly => write!(f, "yearly"),
            TickBehavior::Finished => write!(f, "final"),
            TickBehavior::Unspecified => panic!("Unspecified tick behavior"),
        }
    }
}

impl From<uuid::Uuid> for Uuid {
    fn from(uuid: uuid::Uuid) -> Self {
        let (high, low) = uuid.as_u64_pair();
        Self { high, low }
    }
}

impl From<&uuid::Uuid> for Uuid {
    fn from(uuid: &uuid::Uuid) -> Self {
        let (high, low) = uuid.as_u64_pair();
        Self { high, low }
    }
}

impl From<Uuid> for uuid::Uuid {
    fn from(uuid: Uuid) -> Self {
        Self::from_u64_pair(uuid.high, uuid.low)
    }
}

impl From<&Uuid> for uuid::Uuid {
    fn from(uuid: &Uuid) -> Self {
        Self::from_u64_pair(uuid.high, uuid.low)
    }
}
