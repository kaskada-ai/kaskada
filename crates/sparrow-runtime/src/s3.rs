use std::path::Path;

use anyhow::{anyhow, Context};
use aws_sdk_s3::types::ByteStream;
use aws_sdk_s3::{Client, Endpoint};
use futures::stream::StreamExt;
use itertools::Itertools;
use tempfile::TempPath;
use tokio_util::io::StreamReader;
use tonic::transport::Uri;
use tracing::info;

#[derive(Clone, Debug, Hash, PartialEq, Eq)]
pub struct S3Object {
    pub bucket: String,
    pub key: String,
}

impl S3Object {
    pub fn try_from_uri(path: &str) -> anyhow::Result<Self> {
        if let Some(s3_path) = path.strip_prefix("s3://") {
            if let Some((bucket, key)) = s3_path.split_once('/') {
                Ok(S3Object {
                    bucket: bucket.to_owned(),
                    key: key.to_owned(),
                })
            } else {
                Err(anyhow!(
                    "invalid path provided: '{}' missing bucket/key format",
                    path
                ))
            }
        } else {
            Err(anyhow!(
                "invalid path provided: '{}' missing s3:// prefix",
                path
            ))
        }
    }

    pub fn get_formatted_key(&self) -> String {
        format!("s3://{}/{}", self.bucket, self.key)
    }

    /// Returns the relative path of the key.
    ///
    /// example.
    ///   prefix:   "compute/snapshot/"
    ///   self.key: "/compute/snapshot/file/data"
    ///   returns:  "file/data"
    pub fn get_relative_key_path(&self, prefix: &str) -> anyhow::Result<&str> {
        anyhow::ensure!(!prefix.is_empty(), "Expected prefix to be non-empty");
        let prefix = if prefix.ends_with('/') {
            &prefix[0..prefix.len() - 1]
        } else {
            prefix
        };
        let res = self.key.strip_prefix(prefix).with_context(|| {
            format!(
                "Expected key '{}' to start with prefix '{}', but did not",
                self.key, prefix
            )
        })?;
        // Strip the preceding '/'
        Ok(&res[1..res.len()])
    }

    /// Joins the suffix onto the key separated by a slash.
    ///
    /// If the key has a trailing slash, or the suffix has a preceding
    /// slash, this method will ensure the result is separated only by a
    /// single slash.
    ///
    /// ex. self.key   = "prefix"
    ///     suffix     = "suffix"
    ///     result.key = "prefix/suffix"
    #[must_use]
    pub fn join_delimited(&self, suffix: &str) -> Self {
        let mut result = self.clone();
        result.push_delimited(suffix);
        result
    }

    /// Pushes the suffix onto the key separated by a slash.
    ///
    /// If the key has a trailing slash, or the suffix has a preceding
    /// slash, this method will ensure the result is separated only by a
    /// single slash.
    ///
    /// ex. self.key   = "prefix"
    ///     suffix     = "suffix"
    ///     result.key = "prefix/suffix"
    pub fn push_delimited(&mut self, suffix: &str) {
        let existing_delimiter = self.key.ends_with('/');
        let new_delimiter = suffix.starts_with('/');

        if existing_delimiter && new_delimiter {
            // Double delimited. Drop one from the suffix.
            self.key.push_str(&suffix[1..]);
        } else if existing_delimiter || new_delimiter {
            // Single delimited. Push the entire suffix.
            self.key.push_str(suffix);
        } else {
            // No delimiter. Add one.
            self.key.push('/');
            self.key.push_str(suffix);
        }
    }
}

pub fn is_s3_path(path: &str) -> bool {
    path.to_lowercase().starts_with("s3://")
}

/// S3Helper is a wrapper around the S3 client providing useful methods.
///
/// The underlying client contains only an `Arc`, so it should be cheap
/// to clone.
#[derive(Clone, Debug)]
pub struct S3Helper {
    client: Client,
}

impl S3Helper {
    pub async fn new() -> Self {
        let shared_config = aws_config::from_env().load().await;

        // Create the client.
        let endpoint = std::env::var("AWS_ENDPOINT");
        let s3_config: aws_sdk_s3::config::Config = if let Ok(endpoint) = endpoint {
            // Oeverride the endpoint if needed
            let uri: Uri = endpoint.parse().unwrap();
            let endpoint = Endpoint::immutable(uri);

            aws_sdk_s3::config::Builder::from(&shared_config)
                .endpoint_resolver(endpoint)
                .build()
        } else {
            (&shared_config).into()
        };
        info!("S3 Config: {:?}", s3_config);
        let client = Client::from_conf(s3_config);

        Self { client }
    }

    pub async fn upload_tempfile_to_s3(
        &self,
        s3_key: S3Object,
        local_path: TempPath,
    ) -> anyhow::Result<()> {
        self.upload_s3(s3_key, &local_path).await?;

        // Close and remove the temporary file. This causes errors.
        local_path.close()?;
        Ok(())
    }

    pub async fn download_s3(
        &self,
        s3_object: S3Object,
        target_local_path: impl AsRef<Path>,
    ) -> anyhow::Result<()> {
        let key = Some(s3_object.key);
        let bucket = Some(s3_object.bucket);
        let download = self
            .client
            .get_object()
            .set_key(key)
            .set_bucket(bucket)
            .send()
            .await
            .context("Unable to download object from s3")?;

        let mut file = tokio::fs::File::create(target_local_path).await?;
        let mut body = StreamReader::new(
            download
                .body
                // Convert S3 download errors into `io:Error` so we can use `tokio::io::copy`
                .map(|item| item.map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))),
        );
        // https://docs.rs/tokio/1.14.0/tokio/io/fn.copy.html
        tokio::io::copy(&mut body, &mut file).await?;

        Ok(())
    }

    pub async fn upload_s3(
        &self,
        s3_key: S3Object,
        target_local_path: &Path,
    ) -> anyhow::Result<()> {
        let body = ByteStream::from_path(target_local_path)
            .await
            .with_context(|| {
                format!("unable to convert file at {target_local_path:?} to ByteStream")
            })?;

        let s3_key_clone = s3_key.clone();
        self.client
            .put_object()
            .set_key(Some(s3_key.key))
            .set_bucket(Some(s3_key.bucket))
            .set_body(Some(body))
            .send()
            .await
            .with_context(|| {
                format!("unable to upload file at {target_local_path:?} to S3 at {s3_key_clone:?}")
            })?;

        Ok(())
    }

    /// Lists s3 objects with the given prefix in s3.
    ///
    /// Appends a trailing '/' to the `prefix` if it does not exist.
    pub async fn list_prefix_delimited(
        &self,
        bucket: &str,
        prefix: &str,
    ) -> anyhow::Result<Vec<S3Object>> {
        let prefix = if !prefix.is_empty() {
            let c = &prefix[prefix.len() - 1..prefix.len()];
            if c != "/" {
                let mut res = prefix.to_owned();
                res.push('/');
                res
            } else {
                prefix.to_owned()
            }
        } else {
            String::new()
        };

        let list_obj_res = self
            .client
            .list_objects_v2()
            .set_bucket(Some(bucket.to_owned()))
            .set_prefix(Some(prefix))
            .send()
            .await
            .context("Unable to list objects")?;

        if list_obj_res.is_truncated() {
            anyhow::bail!("Unsupported: did not expect result to be truncated!")
        }

        let res = list_obj_res
            .contents()
            .unwrap_or_default()
            .iter()
            .map(|i| -> anyhow::Result<_> {
                Ok(S3Object {
                    bucket: bucket.to_owned(),
                    key: i.key().context("Could not parse s3 key")?.to_owned(),
                })
            })
            .try_collect()?;

        Ok(res)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_try_from_uri_valid() {
        let valid_uri = "s3://bucket/some/key/path";
        let result = S3Object::try_from_uri(valid_uri).unwrap();
        assert_eq!(result.key, "some/key/path");
        assert_eq!(result.bucket, "bucket");
    }

    #[test]
    fn test_try_from_uri_invalid() {
        let invalid_uri = "s3:/bucket/some/key/path";
        assert_eq!(
            S3Object::try_from_uri(invalid_uri).unwrap_err().to_string(),
            "invalid path provided: 's3:/bucket/some/key/path' missing s3:// prefix",
        );
    }

    #[test]
    fn test_get_formatted_key() {
        let expected = "s3://bucket/some/key/path";
        let obj = S3Object {
            key: "some/key/path".to_owned(),
            bucket: "bucket".to_owned(),
        };
        assert_eq!(obj.get_formatted_key(), expected);
    }

    #[test]
    fn test_get_relative_key_path() {
        let obj = S3Object {
            key: "some/key/path".to_owned(),
            bucket: "bucket".to_owned(),
        };
        let relative_path = obj.get_relative_key_path("some/key").unwrap();
        assert_eq!(relative_path, "path");

        // With trailing slash
        // Regression test against splitting the key incorrectly.
        let relative_path = obj.get_relative_key_path("some/key/").unwrap();
        assert_eq!(relative_path, "path");
    }

    #[test]
    fn test_get_relative_key_path_fails_invalid_prefix() {
        let obj = S3Object {
            key: "some/key/path".to_owned(),
            bucket: "bucket".to_owned(),
        };

        assert_eq!(
            obj.get_relative_key_path("bogus").unwrap_err().to_string(),
            "Expected key 'some/key/path' to start with prefix 'bogus', but did not"
        );
    }

    #[test]
    fn test_join_delimited_adds_delimiter() {
        let obj = S3Object {
            key: "some/key/path".to_owned(),
            bucket: "bucket".to_owned(),
        };
        let res = obj.join_delimited("suffix");
        assert_eq!(res.key, "some/key/path/suffix");
    }

    #[test]
    fn test_join_delimited_with_existing_delimiters() {
        let obj_key_trailing_slash = S3Object {
            key: "some/key/path/".to_owned(),
            bucket: "bucket".to_owned(),
        };
        let res = obj_key_trailing_slash.join_delimited("suffix");
        assert_eq!(res.key, "some/key/path/suffix");

        let res = obj_key_trailing_slash.join_delimited("/suffix");
        assert_eq!(res.key, "some/key/path/suffix");
    }

    #[test]
    fn test_is_s3_path_valid() {
        assert!(is_s3_path("s3://bucket/key"));
        assert!(!is_s3_path("s3//not/valid"));
        assert!(!is_s3_path("s3:/not/valid"));
        assert!(!is_s3_path("bucket/not/valid"));
    }
}
