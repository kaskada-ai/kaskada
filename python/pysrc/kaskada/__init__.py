"""Kaskada query builder and local execution engine."""
from __future__ import annotations

from . import destinations, plot, results, sources, windows
from ._execution import Execution, ResultIterator
from ._session import init_session
from ._timestream import Arg, LiteralValue, Timestream, record
from .udf import udf


__all__ = [
    "Arg",
    "destinations",
    "Execution",
    "init_session",
    "LiteralValue",
    "plot",
    "record",
    "ResultIterator",
    "results",
    "sources",
    "Timestream",
    "udf",
    "windows",
]
