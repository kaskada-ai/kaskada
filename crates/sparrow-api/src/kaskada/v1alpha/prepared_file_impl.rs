use chrono::NaiveDateTime;

use crate::kaskada::v1alpha::PreparedFile;

#[derive(derive_more::Display, Debug)]
pub enum Error {
    #[display(fmt = "missing field '{field}")]
    MissingField { field: &'static str },
    #[display(fmt = "invalid timestamp {timestamp:?}")]
    InvalidTimestamp {
        timestamp: prost_wkt_types::Timestamp,
    },
}

impl error_stack::Context for Error {}

impl From<&PreparedFile> for PreparedFile {
    fn from(file: &PreparedFile) -> Self {
        file.clone()
    }
}

impl PreparedFile {
    pub fn min_event_time(&self) -> error_stack::Result<NaiveDateTime, Error> {
        let Some(timestamp) = self.min_event_time.as_ref() else {
            error_stack::bail!(Error::MissingField {
                field: "min_event_time"
            })
        };
        NaiveDateTime::from_timestamp_opt(timestamp.seconds, timestamp.nanos as u32).ok_or_else(
            || {
                error_stack::report!(Error::InvalidTimestamp {
                    timestamp: timestamp.clone()
                })
            },
        )
    }

    pub fn max_event_time(&self) -> error_stack::Result<NaiveDateTime, Error> {
        let Some(timestamp) = self.max_event_time.as_ref() else {
            error_stack::bail!(Error::MissingField {
                field: "max_event_time"
            })
        };
        NaiveDateTime::from_timestamp_opt(timestamp.seconds, timestamp.nanos as u32).ok_or_else(
            || {
                error_stack::report!(Error::InvalidTimestamp {
                    timestamp: timestamp.clone()
                })
            },
        )
    }
}
