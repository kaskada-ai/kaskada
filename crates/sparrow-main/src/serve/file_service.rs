use anyhow::Context;
use sparrow_api::kaskada::v1alpha::file_service_server::FileService;
use sparrow_api::kaskada::v1alpha::get_metadata_request::Source;
use sparrow_api::kaskada::v1alpha::Schema;
use sparrow_api::kaskada::v1alpha::{
    file_path, FilePath, GetMetadataRequest, GetMetadataResponse, MergeMetadataRequest,
    MergeMetadataResponse, SourceMetadata,
};
use sparrow_core::context_code;
use sparrow_runtime::s3::{is_s3_path, S3Helper, S3Object};
use sparrow_runtime::RawMetadata;
use tempfile::NamedTempFile;
use tonic::{Code, Response};

use crate::serve::error_status::IntoStatus;

#[derive(Debug)]
pub(super) struct FileServiceImpl {
    s3: S3Helper,
}

impl FileServiceImpl {
    pub fn new(s3: S3Helper) -> Self {
        Self { s3 }
    }
}

#[tonic::async_trait]
impl FileService for FileServiceImpl {
    #[tracing::instrument]
    async fn get_metadata(
        &self,
        request: tonic::Request<GetMetadataRequest>,
    ) -> Result<tonic::Response<GetMetadataResponse>, tonic::Status> {
        let s3 = self.s3.clone();

        match tokio::spawn(get_metadata(s3, request)).await {
            Ok(result) => result.into_status(),
            Err(panic) => {
                tracing::error!("Panic during prepare: {panic}");
                Err(tonic::Status::internal("panic during prepare"))
            }
        }
    }

    #[tracing::instrument]
    async fn merge_metadata(
        &self,
        _request: tonic::Request<MergeMetadataRequest>,
    ) -> Result<tonic::Response<MergeMetadataResponse>, tonic::Status> {
        Err(tonic::Status::unimplemented(
            "merge_metadata not implemented",
        ))
    }
}

async fn get_metadata(
    s3: S3Helper,
    request: tonic::Request<GetMetadataRequest>,
) -> anyhow::Result<tonic::Response<GetMetadataResponse>> {
    let request = request.into_inner();
    let source = request.source.context("source")?;
    let source: FilePath = match source {
        Source::FilePath(fp) => fp,
        Source::PulsarSource(_) => anyhow::bail!("Pulsar source unsupported"),
    };
    let source_metadata = get_source_metadata(&s3, &source).await?;

    Ok(Response::new(GetMetadataResponse {
        source_metadata: Some(source_metadata),
    }))
}

pub(crate) async fn get_source_metadata(
    s3: &S3Helper,
    source: &FilePath,
) -> anyhow::Result<SourceMetadata> {
    let source = source
        .path
        .as_ref()
        .context(context_code!(Code::InvalidArgument, "Missing source_path"))?;

    let download_file = NamedTempFile::new()?;
    let download_file_path = download_file.into_temp_path();
    let source = match source {
        file_path::Path::ParquetPath(path) => {
            if is_s3_path(path) {
                let s3_object = S3Object::try_from_uri(path)?;
                s3.download_s3(s3_object, download_file_path.to_owned())
                    .await?;
                file_path::Path::ParquetPath(download_file_path.to_string_lossy().to_string())
            } else {
                file_path::Path::ParquetPath(path.to_string())
            }
        }
        file_path::Path::CsvPath(path) => {
            if is_s3_path(path) {
                let s3_object = S3Object::try_from_uri(path)?;
                s3.download_s3(s3_object, download_file_path.to_owned())
                    .await?;
                file_path::Path::CsvPath(download_file_path.to_string_lossy().to_string())
            } else {
                file_path::Path::CsvPath(path.to_string())
            }
        }
        file_path::Path::CsvData(data) => file_path::Path::CsvData(data.to_string()),
    };

    let metadata = RawMetadata::try_from(&source)?;
    let schema = Schema::try_from(metadata.table_schema.as_ref()).with_context(|| {
        format!(
            "Unable to encode schema {:?} for source file {:?}",
            metadata.table_schema, source
        )
    })?;

    Ok(SourceMetadata {
        schema: Some(schema),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_timestamp_with_timezone_gets_metadata() {
        // Regression test for handling files with a timestamp with a configured
        // timezone.
        let path = sparrow_testing::testdata_path("eventdata/sample_event_data.parquet");

        let s3_helper = S3Helper::new().await;
        let file_service = FileServiceImpl::new(s3_helper);

        // TODO: We may be able to clean this up to test the metadata logic
        // without creating the S3 helper. But for now, this will suffice.
        let result = file_service
            .get_metadata(tonic::Request::new(GetMetadataRequest {
                source: Some(Source::FilePath(FilePath {
                    path: Some(file_path::Path::ParquetPath(
                        path.canonicalize().unwrap().to_string_lossy().to_string(),
                    )),
                })),
            }))
            .await
            .unwrap()
            .into_inner();

        insta::assert_yaml_snapshot!(result);
    }

    #[tokio::test]
    async fn test_get_metadata_csv_path() {
        let path =
            sparrow_testing::testdata_path("eventdata/2c889258-d676-4922-9a92-d7e9c60c1dde.csv");

        let s3_helper = S3Helper::new().await;
        let file_service = FileServiceImpl::new(s3_helper);

        // TODO: We may be able to clean this up to test the metadata logic
        // without creating the S3 helper. But for now, this will suffice.
        let result = file_service
            .get_metadata(tonic::Request::new(GetMetadataRequest {
                source: Some(Source::FilePath(FilePath {
                    path: Some(file_path::Path::CsvPath(
                        path.canonicalize().unwrap().to_string_lossy().to_string(),
                    )),
                })),
            }))
            .await
            .unwrap()
            .into_inner();

        insta::assert_yaml_snapshot!(result);
    }

    #[tokio::test]
    async fn test_get_metadata_csv_data() {
        let csv_data = "id,name,value,value2\n1,awkward,taco,123\n2,taco,awkward,456\n";
        let s3_helper = S3Helper::new().await;
        let file_service = FileServiceImpl::new(s3_helper);
        let result = file_service
            .get_metadata(tonic::Request::new(GetMetadataRequest {
                source: Some(Source::FilePath(FilePath {
                    path: Some(file_path::Path::CsvData(csv_data.to_string())),
                })),
            }))
            .await
            .unwrap()
            .into_inner();

        insta::assert_yaml_snapshot!(result);
    }
}
