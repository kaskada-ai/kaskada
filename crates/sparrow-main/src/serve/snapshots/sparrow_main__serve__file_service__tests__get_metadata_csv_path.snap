---
source: crates/sparrow-main/src/serve/file_service.rs
expression: result
---
source_metadata:
  schema:
    fields:
      - name: ""
        data_type:
          kind:
            Primitive: 6
      - name: anonymousId
        data_type:
          kind:
            Primitive: 14
      - name: projectId
        data_type:
          kind:
            Primitive: 14
      - name: type
        data_type:
          kind:
            Primitive: 14
      - name: timestamp
        data_type:
          kind:
            Primitive: 14
      - name: sentAt
        data_type:
          kind:
            Primitive: 14
      - name: receivedAt
        data_type:
          kind:
            Primitive: 14
      - name: channel
        data_type:
          kind:
            Primitive: 14
      - name: event
        data_type:
          kind:
            Primitive: 14
      - name: messageId
        data_type:
          kind:
            Primitive: 14
      - name: name
        data_type:
          kind:
            Primitive: 14
      - name: userId
        data_type:
          kind:
            Primitive: 14
      - name: version
        data_type:
          kind:
            Primitive: 6
      - name: context_campaign_content
        data_type:
          kind:
            Primitive: 14
      - name: context_campaign_medium
        data_type:
          kind:
            Primitive: 14
      - name: context_campaign_name
        data_type:
          kind:
            Primitive: 14
      - name: context_campaign_source
        data_type:
          kind:
            Primitive: 14
      - name: context_campaign_term
        data_type:
          kind:
            Primitive: 14
      - name: context_ip
        data_type:
          kind:
            Primitive: 14
      - name: context_library_name
        data_type:
          kind:
            Primitive: 14
      - name: context_library_version
        data_type:
          kind:
            Primitive: 14
      - name: context_locale
        data_type:
          kind:
            Primitive: 14
      - name: context_page_path
        data_type:
          kind:
            Primitive: 14
      - name: context_page_referrer
        data_type:
          kind:
            Primitive: 14
      - name: context_page_search
        data_type:
          kind:
            Primitive: 14
      - name: context_page_title
        data_type:
          kind:
            Primitive: 14
      - name: context_page_url
        data_type:
          kind:
            Primitive: 14
      - name: context_userAgent
        data_type:
          kind:
            Primitive: 14
      - name: properties_formula
        data_type:
          kind:
            Primitive: 14
      - name: properties_isTarget
        data_type:
          kind:
            Primitive: 14
      - name: properties_name
        data_type:
          kind:
            Primitive: 14
      - name: properties_path
        data_type:
          kind:
            Primitive: 14
      - name: properties_referrer
        data_type:
          kind:
            Primitive: 14
      - name: properties_search
        data_type:
          kind:
            Primitive: 14
      - name: properties_selected
        data_type:
          kind:
            Primitive: 14
      - name: properties_title
        data_type:
          kind:
            Primitive: 14
      - name: properties_type
        data_type:
          kind:
            Primitive: 14
      - name: properties_url
        data_type:
          kind:
            Primitive: 14
      - name: traits_activeAccountId
        data_type:
          kind:
            Primitive: 14
      - name: traits_BETA_DATA_SCIENCE
        data_type:
          kind:
            Primitive: 14
      - name: traits_company
        data_type:
          kind:
            Primitive: 14
      - name: traits_DSE_FREE
        data_type:
          kind:
            Primitive: 14
      - name: traits_email
        data_type:
          kind:
            Primitive: 14
      - name: traits_firstName
        data_type:
          kind:
            Primitive: 14
      - name: traits_job
        data_type:
          kind:
            Primitive: 14
      - name: traits_lastName
        data_type:
          kind:
            Primitive: 14
      - name: traits_name
        data_type:
          kind:
            Primitive: 14
      - name: traits_PAID_DS_ACCOUNT
        data_type:
          kind:
            Primitive: 14
      - name: traits_PUBLIC
        data_type:
          kind:
            Primitive: 14
      - name: traits_userId
        data_type:
          kind:
            Primitive: 14
      - name: subsort_id
        data_type:
          kind:
            Primitive: 6

